# Week 4 Workshop: Data Filtering & Selection
## Traffic Accident Vehicles in Colombia

Practice filtering techniques to answer analytical questions about vehicles involved in traffic accidents.

**Duration:** ~2 hours

**Objectives:**
- Filter rows using comparison operators
- Combine conditions with & (AND), | (OR), ~ (NOT)
- Use .isin(), .between(), .str.contains()
- Use .query() for clean syntax
- Translate analytical questions into filter code

---

## Setup
Run this cell to load the dataset and see the overview.

In [1]:
import pandas as pd
import numpy as np

# Load the Traffic Accident Vehicles dataset
df = pd.read_csv('../data/vehiculos_accidentes.csv')

# Quick overview
print(f"Dataset: {df.shape[0]} rows x {df.shape[1]} columns")
print(f"\nColumns: {df.columns.tolist()}")
print(f"\nVehicle types: {df['tipo_vehiculo'].nunique()} unique")
print(f"Departments: {df['departamento_accidente'].nunique()} unique")
print(f"Severity levels: {df['gravedad_accidente'].unique().tolist()}")
print(f"Brands: {df['marca_vehiculo'].nunique()} unique")
print(f"\nTop 5 vehicle types:")
print(df['tipo_vehiculo'].value_counts().head())
df.head(5)

Dataset: 20000 rows x 9 columns

Columns: ['marca_vehiculo', 'modelo_vehiculo', 'tipo_vehiculo', 'edad_vehiculo', 'fecha_accidente', 'gravedad_accidente', 'departamento_accidente', 'municipio_accidente', 'autoridad_de_transito']

Vehicle types: 16 unique
Departments: 29 unique
Severity levels: ['CON HERIDOS', 'CON MUERTOS']
Brands: 159 unique

Top 5 vehicle types:
tipo_vehiculo
MOTOCICLETA    11301
AUTOMOVIL       4185
CAMIONETA       1777
BUS              840
CAMION           618
Name: count, dtype: int64


,marca_vehiculo,modelo_vehiculo,tipo_vehiculo,edad_vehiculo,fecha_accidente,gravedad_accidente,departamento_accidente,municipio_accidente,autoridad_de_transito
0,AKT,2026,MOTOCICLETA,1.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
1,BAJAJ,2022,MOTOCICLETA,4.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
2,RENAULT,2021,AUTOMOVIL,5.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
3,BAJAJ,2017,MOTOCICLETA,10.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
4,BAJAJ,2014,MOTOCICLETA,11.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA


---

## Part 1: Single Condition Filters (20 minutes)

Practice the basic filtering pattern: `df[df['column'] operator value]`

### Task 1.1: Filter by vehicle type
Get all records where the vehicle is a **MOTOCICLETA** (motorcycle).

Print the number of rows and display the first 5.

In [35]:
# YOUR CODE HERE
# Hint: df[df['tipo_vehiculo'] == 'MOTOCICLETA']
df_filtro1 = df[df['tipo_vehiculo'] == 'MOTOCICLETA']
print(f"Total motorcycle accidents: {df_filtro1.shape[0]}")
print(f"porcentaje: {len(df_filtro1) / len(df) * 100:.1f}%")
df_filtro1.head(5)




Total motorcycle accidents: 11301
porcentaje: 56.5%


,marca_vehiculo,modelo_vehiculo,tipo_vehiculo,edad_vehiculo,fecha_accidente,gravedad_accidente,departamento_accidente,municipio_accidente,autoridad_de_transito
0,AKT,2026,MOTOCICLETA,1.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
1,BAJAJ,2022,MOTOCICLETA,4.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
3,BAJAJ,2017,MOTOCICLETA,10.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
4,BAJAJ,2014,MOTOCICLETA,11.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
5,BAJAJ,2011,MOTOCICLETA,15.0,12/2025,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA


### Task 1.2: Filter by comparison
Find vehicles with age greater than 15 years (`edad_vehiculo > 15`).

**Question:** How many old vehicles were involved in accidents? What are the most common brands among them?

In [ ]:
# YOUR CODE HERE
# Hint: after filtering, try result['marca_vehiculo'].value_counts().head() to see the top brands
df_filtro2 = df[df['edad_vehiculo'] > 15]
print(f"Total accidents with vehicles older than 15 years: {df_filtro2.shape[0]}")
df_filtro2['marca_vehiculo'].value_counts().head()


Total accidents with vehicles older than 15 years: 3369


marca_vehiculo
CHEVROLET    864
HYUNDAI      336
RENAULT      243
YAMAHA       195
SUZUKI       190
Name: count, dtype: int64

### Task 1.3: Filter by exact text
Get all accidents that occurred in the department **ATLANTICO**.

Sort the results by `fecha_accidente` and display the first 10 rows.

In [9]:
# YOUR CODE HERE
# Hint: after filtering, use .sort_values('fecha_accidente') and .head(10)
df_filtro3 = df[df['departamento_accidente'] == 'ATLANTICO']
print(f"Total accidents in Antioquia: {df_filtro3.shape[0]}")
df_filtro3.sort_values('fecha_accidente').head(10)


Total accidents in Antioquia: 867


,marca_vehiculo,modelo_vehiculo,tipo_vehiculo,edad_vehiculo,fecha_accidente,gravedad_accidente,departamento_accidente,municipio_accidente,autoridad_de_transito
19948,AKT,2016,MOTOCICLETA,10.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19939,AKT,2019,MOTOCICLETA,7.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19940,BAJAJ,2010,MOTOCICLETA,16.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19941,BAJAJ,2022,MOTOCICLETA,4.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19942,BAJAJ,2023,MOTOCICLETA,4.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19947,AKT,2023,MOTOCICLETA,3.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19944,TOYOTA,2020,CAMIONETA,6.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19945,RENAULT,2007,AUTOMOVIL,19.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19946,AKT,2023,MOTOCICLETA,4.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA
19943,RENAULT,2022,AUTOMOVIL,4.0,12/2022,CON HERIDOS,ATLANTICO,BARRANQUILLA,STRIA DTAL TTO BARRANQUILLA


### Task 1.4: Filter with not equal
Get all records where the severity is NOT "CON HERIDOS" (i.e., get only the fatal accidents).

Print the count and compare it to the total dataset size.

In [11]:
# YOUR CODE HERE
# print(f"Total records: {len(df)}")
# fatal = ...
# print(f"Fatal accidents: {len(fatal)}")
# print(f"Percentage: {len(fatal) / len(df) * 100:.1f}%")

df_filtro4 = df[df['gravedad_accidente'] != 'CON HERIDOS']
print(f"Total records: {len(df)}")
print(f"Fatal accidents: {len(df_filtro4)}")
print(f"Percentage: {len(df_filtro4) / len(df) * 100:.1f}%")


Total records: 20000
Fatal accidents: 1352
Percentage: 6.8%


---

## Part 2: Combining Conditions (30 minutes)

Combine multiple conditions to answer more specific questions.

**Rules:**
- Use `&` (AND), `|` (OR), `~` (NOT) - NOT `and`, `or`, `not`
- Always wrap each condition in parentheses

### Task 2.1: Two AND conditions
Find motorcycles (`MOTOCICLETA`) involved in fatal accidents (`CON MUERTOS`).

**Question:** How many fatal motorcycle accidents are in the dataset?

In [16]:
# YOUR CODE HERE
# Pattern: df[(condition1) & (condition2)]
df_filtro5 = df[(df['tipo_vehiculo'] == 'MOTOCICLETA') & (df['gravedad_accidente'] == 'CON MUERTOS')]
print(f"Total fatal motorcycle accidents: {df_filtro5.shape[0]}")   

Total fatal motorcycle accidents: 758


### Task 2.2: Three AND conditions
Find **AUTOMOVIL** vehicles, in **BOGOTA D.C.**, with severity **CON HERIDOS**.

Show the brand (`marca_vehiculo`) and model year (`modelo_vehiculo`) columns of the results.

In [23]:
# YOUR CODE HERE


# Hint: after filtering, select columns with result[['marca_vehiculo', 'modelo_vehiculo']]
df_filtro6 = df[(df['tipo_vehiculo'] == 'AUTOMOVIL') & (df['gravedad_accidente'] == 'CON HERIDOS') & (df['departamento_accidente'] == 'BOGOTADC')]
df_filtro6[['marca_vehiculo', 'modelo_vehiculo']]

,marca_vehiculo,modelo_vehiculo
7357,RENAULT,2007
7375,RENAULT,2023
7383,CITROEN,1994
7406,RENAULT,2007
7439,DAEWOO,2000
...,...,...
13457,RENAULT,2018
13468,RENAULT,2012
13628,RENAULT,2008
13630,RENAULT,2021


### Task 2.3: OR and .isin()
Find all accidents involving either **CHEVROLET**, **TOYOTA**, or **MAZDA** vehicles.

Try both approaches:
1. Using `|` (OR) operators
2. Using `.isin()`

Verify both give the same number of rows.

In [24]:
# Approach 1: Using | (OR)
filtro_or = df[
    (df['marca_vehiculo'] == 'CHEVROLET') |
    (df['marca_vehiculo'] == 'TOYOTA') |
    (df['marca_vehiculo'] == 'MAZDA')
]

# Approach 2: Using .isin()
filtro_isin = df[
    df['marca_vehiculo'].isin(['CHEVROLET','TOYOTA','MAZDA'])
]

# Verify both have the same number of rows
print("OR:", len(filtro_or))
print("isin:", len(filtro_isin))


OR: 3104
isin: 3104


### Task 2.4: Mixed AND + OR
Find accidents in **ANTIOQUIA** or **VALLE DEL CAUCA** where the vehicle is older than 10 years.

**Hint:** Combine `.isin()` with another condition:
```python
df[(df['col'].isin([...]) & (df['other_col'] > value)]
```

In [26]:
# YOUR CODE HERE
resultado = df[
    (df['departamento_accidente'].isin(['ANTIOQUIA','VALLE DEL CAUCA'])) &
    (df['edad_vehiculo'] > 10)
]

resultado.head()


,marca_vehiculo,modelo_vehiculo,tipo_vehiculo,edad_vehiculo,fecha_accidente,gravedad_accidente,departamento_accidente,municipio_accidente,autoridad_de_transito
68,BAJAJ,2010,MOTOCICLETA,17.0,12/2025,CON HERIDOS,ANTIOQUIA,APARTADO,STRIA MCPAL TTEyTTO APARTADO
69,YAMAHA,2016,MOTOCICLETA,11.0,12/2025,CON HERIDOS,ANTIOQUIA,APARTADO,STRIA MCPAL TTEyTTO APARTADO
71,YAMAHA,2014,MOTOCICLETA,12.0,12/2025,CON HERIDOS,ANTIOQUIA,APARTADO,STRIA MCPAL TTEyTTO APARTADO
117,AKT,2012,MOTOCICLETA,14.0,12/2025,CON HERIDOS,ANTIOQUIA,LA CEJA - SAN JOSÉ,DIRECCION DE TTOYTTE POLICIA NACIONAL - DITRA
123,BAJAJ,2014,MOTOCICLETA,12.0,12/2025,CON HERIDOS,ANTIOQUIA,RIONEGRO,SDT SUBSECRETARÍA DE MOVILIDAD RIONEGRO


### Task 2.5: NOT with .isin()
Get all accident records EXCLUDING motorcycles (`MOTOCICLETA`) and buses (`BUS`).

**Pattern:** `df[~df['col'].isin([list])]`

Print the count and the remaining vehicle types.

In [27]:
# YOUR CODE HERE
# After filtering, try: result['tipo_vehiculo'].value_counts() to see remaining types

resultado = df[
    ~df['tipo_vehiculo'].isin(['MOTOCICLETA','BUS'])
]

print("Cantidad:", len(resultado))

print("\nTipos restantes:")
print(resultado['tipo_vehiculo'].value_counts())


Cantidad: 7859

Tipos restantes:
tipo_vehiculo
AUTOMOVIL        4185
CAMIONETA        1777
CAMION            618
CAMPERO           496
TRACTOCAMION      292
MICROBUS          208
BUSETA            135
VOLQUETA          102
MOTOCARRO          38
MONTACARGAS         2
CICLOMOTOR          2
TRACTOR             2
CUATRIMOTO          1
MAQ. AGRICOLA       1
Name: count, dtype: int64


---

## Part 3: Convenience Methods (20 minutes)

Practice .between(), .str.contains(), .str.startswith(), and .query() for common filtering patterns.

### Task 3.1: .between() for model year
Find vehicles with model year between **2015 and 2020** (inclusive).

**Pattern:** `df[df['col'].between(low, high)]`

Print how many records match and show the distribution of model years in the result.

In [28]:
# YOUR CODE HERE
# Hint: after filtering, use result['modelo_vehiculo'].value_counts().sort_index()
resultado = df[df['modelo_vehiculo'].between(2015, 2020)]

print("Cantidad:", len(resultado))

print(resultado['modelo_vehiculo'].value_counts().sort_index())


Cantidad: 5999
modelo_vehiculo
2015    1026
2016    1003
2017     923
2018     863
2019    1022
2020    1162
Name: count, dtype: int64


### Task 3.2: .between() for age range
Get vehicles aged **0 to 5 years** (relatively new vehicles).

Print the count and the percentage of the total dataset.

In [ ]:
# YOUR CODE HERE
resultado = df[df['edad_vehiculo'].between(0,5)]

print("Cantidad:", len(resultado))

print("Porcentaje:", len(resultado)/len(df)*100, "%")

### Task 3.3: .str.contains() for text search
Find transit authorities (`autoridad_de_transito`) whose name contains **"BOGOTA"**.

Print the unique authority names that match.

**Remember:** Always add `na=False` to handle potential NaN values.

In [ ]:
# YOUR CODE HERE
# Pattern: df[df['col'].str.contains('text', na=False)]
resultado = df[
    df['autoridad_de_transito'].str.contains('BOGOTA', na=False)
]

print(resultado['autoridad_de_transito'].unique())


### Task 3.4: .str.startswith()
Find vehicle brands that start with **"CH"** (Chevrolet, Changan, Chery, etc.).

Print the unique brand names that match and the total number of records.

**Pattern:** `df[df['col'].str.startswith('text', na=False)]`

In [ ]:
# YOUR CODE HERE
resultado = df[
    df['marca_vehiculo'].str.startswith('CH', na=False)
]

print("Marcas:")
print(resultado['marca_vehiculo'].unique())

print("\nCantidad:", len(resultado))

### Task 3.5: .query() method
Rewrite this boolean indexing filter using `.query()`:

```python
df[(df['edad_vehiculo'] > 10) & (df['departamento_accidente'] == 'ANTIOQUIA') & (df['tipo_vehiculo'] == 'MOTOCICLETA')]
```

Verify both approaches give the same number of rows.

In [29]:
# Boolean indexing version (given):
result_bool = df[
    (df['edad_vehiculo'] > 10) &
    (df['departamento_accidente'] == 'ANTIOQUIA') &
    (df['tipo_vehiculo'] == 'MOTOCICLETA')
]

# YOUR CODE HERE - .query() version:
# result_query = df.query('...')

# Verify:
# print(f"Boolean indexing: {len(result_bool)} rows")
# print(f".query(): {len(result_query)} rows")

result_query = df.query(
    "edad_vehiculo > 10 and departamento_accidente == 'ANTIOQUIA' and tipo_vehiculo == 'MOTOCICLETA'"
)

print(f"Boolean indexing: {len(result_bool)} rows")
print(f".query(): {len(result_query)} rows")


Boolean indexing: 999 rows
.query(): 999 rows


---

## Part 4: Analytical Questions (30 minutes)

Now use your filtering skills to answer real questions about traffic safety in Colombia. For each question:
1. Write the filter
2. Display relevant columns or value counts
3. Write a 1-2 sentence interpretation of the result

### Question 4.1: Fatal accidents by vehicle type
Which vehicle type is most involved in **fatal accidents** (`CON MUERTOS`)?

Filter for fatal accidents and count by `tipo_vehiculo`. Sort from highest to lowest.

In [2]:
# YOUR CODE HERE
# Hint: filter first, then use ['tipo_vehiculo'].value_counts()
fatal = df[df['gravedad_accidente'] == 'CON MUERTOS']

conteo = fatal['tipo_vehiculo'].value_counts()

print(conteo)

tipo_vehiculo
MOTOCICLETA      758
AUTOMOVIL        154
CAMIONETA        118
TRACTOCAMION     108
CAMION           103
BUS               54
CAMPERO           20
VOLQUETA          14
MICROBUS          13
BUSETA             9
MAQ. AGRICOLA      1
Name: count, dtype: int64


**Your interpretation:** *(Write 1-2 sentences about what you observe)*


### Question 4.2: Motorcycle accidents by department
Which departments have the most **motorcycle** accidents?

Filter for `MOTOCICLETA` and count by `departamento_accidente`. Show the top 10.

In [31]:
# YOUR CODE HERE
motos = df[df['tipo_vehiculo'] == 'MOTOCICLETA']

print(motos['departamento_accidente'].value_counts().head(10))

departamento_accidente
ANTIOQUIA        3992
BOGOTADC         1599
VALLEDELCAUCA    1568
SANTANDER         562
ATLANTICO         387
TOLIMA            329
CUNDINAMARCA      320
CALDAS            319
RISARALDA         319
META              311
Name: count, dtype: int64


**Your interpretation:** *(Write 1-2 sentences)*


### Question 4.3: Vehicle age and severity
Are older vehicles more likely to be in **severe** (fatal) accidents?

Calculate the average `edad_vehiculo` for each `gravedad_accidente` level.

**Hint:** Filter for each severity level separately and compute `.mean()`, or use `df.groupby('gravedad_accidente')['edad_vehiculo'].mean()`.

In [3]:
# YOUR CODE HERE
print(df.groupby('gravedad_accidente')['edad_vehiculo'].mean())

gravedad_accidente
CON HERIDOS     9.642999
CON MUERTOS    10.815089
Name: edad_vehiculo, dtype: float64


**Your interpretation:** *(Write 1-2 sentences)*


### Question 4.4: Your own question
Design and answer your own analytical question about the traffic accident data.

Ideas:
- Which brands are most involved in fatal accidents in a specific department?
- How do accident patterns differ between Medellin and Bogota?
- What is the age profile of motorcycles vs. cars in accidents?
- Which municipalities have the most accidents with heavy vehicles (CAMION, TRACTOCAMION)?

**Your question:** *(Write it here)*


In [33]:
# YOUR CODE HERE
fatal = df[df['gravedad_accidente'] == 'CON MUERTOS']

print(fatal['marca_vehiculo'].value_counts().head(10))

marca_vehiculo
BAJAJ            195
YAMAHA           146
CHEVROLET        129
AKT              105
HONDA             98
SUZUKI            90
KENWORTH          55
RENAULT           47
INTERNATIONAL     40
TVS               37
Name: count, dtype: int64


**Your interpretation:** *(Write 1-2 sentences)*


---

## Part 5: Reflection

Answer these questions about your experience with data filtering.

### Reflection 1
Which filtering method did you find most useful: boolean indexing, .isin(), .between(), .str.contains(), or .query()? Why?

**Your answer:**

The most useful method depends on the context: for simple filters with operators `&`, `|`, `~`, it's clear and straightforward. However, `.isin()` is especially useful when you need to filter by multiple values ​​in a list (such as specific vehicle makes), because it's more readable than writing multiple OR conditions. `.between()` is ideal for numeric ranges (age, vehicle model) and is an elegant way to express bound inclusion. `.query()` is excellent for complex queries because the syntax is more similar to natural language, which improves code readability.

### Reflection 2
What questions do you want to answer about **YOUR** project dataset using filtering? List at least 3 questions.

**Your answer:**
1. What type of vehicle is most frequently involved in fatal accidents, and in which department do these accidents occur most often? (As we observed, motorcycles account for 758 of 1,352 fatal accidents—56% of the total.)

2. Is there a relationship between vehicle age and accident severity? Are older vehicles more likely to be involved in fatal accidents? (I observed that the average age in fatal accidents is 10.8 years versus 9.6 years in accidents with injuries.)

3. What are the most frequent vehicle makes involved in fatal accidents in cities like Bogotá and Medellín, and what safety patterns can we identify?

### Reflection 3
Why is it important to clean the data (Week 3) BEFORE filtering it (Week 4)? What problems could arise if you filter dirty data?

**Your answer:**


1. **Inconsistencies in values**: If department names vary (e.g., 'BOGOTA', 'BOGOTA D.C.', 'Bogota'), your filters may miss records. This would lead to incomplete analysis and biased conclusions about accidents in those regions.

2. **NULL/NaN values**: If you have missing values ​​in filter columns (such as `accident_department`), records may be inadvertently excluded even if they are relevant, resulting in data loss.

3. **Incorrect data types**: If `vehicle_age` is loaded as text instead of a number, the `vehicle_age > 15` comparison would fail or be incorrect, affecting analyses of older vehicles.

4. **Duplicates**: Duplicate records would inflate counts and statistics, making numbers like "758 motorcycles in fatal accidents" misleading.

5. **Special characters and spaces**: Values ​​like 'MOTORCYCLE' (with a space) would not match 'MOTORCYCLE' in your filter, fragmenting your results.

---

## Summary

In this workshop you practiced:

| Skill | Methods |
|-------|--------|
| Single conditions | `==`, `!=`, `>`, `<`, `>=`, `<=` |
| Combining conditions | `&` (AND), `|` (OR), `~` (NOT) |
| Match a list | `.isin([list])` |
| Numeric range | `.between(low, high)` |
| Text pattern | `.str.contains()`, `.str.startswith()` |
| Clean syntax | `.query('expression')` |

These skills are essential for Milestone 1: you need to filter your project dataset to focus on specific subsets of data for analysis.

**Next week:** Exploratory Data Analysis (EDA) - using GroupBy and statistics to discover patterns.

---

*Week 4 - Data Analytics Course - Universidad Cooperativa de Colombia*